### DISCUSSION TITLE
How to force source HTTP_listener to HTTP code 500 when exception raised in a transformer ?

### DISCUSSION THREAD

**Discussion by cspielmann** [nextgenhealthcare/connect#5731] [Q&A]
Hello,
As far as I can see, when an exception is raised into a transformer javascript, then the HTTP_listener source responds with HTTP code 200.
I would like to force it to respond with an error 500.
Maybe in the postprocessor, i can force a response status in the response map.
But How can I trap the exception. I don't want to trap in every javascripts of my channel.

Any idea ?
thanks a lot.
regards.
Christophe

**Comment by jonbartels** [comment]
Is the response code in your source connector hardcoded to 200?

If you leave it blank Mirth should return a 500 when your connector errors.

**Reply by cspielmann** [reply]
Currently, it is left blank. 
I thought initially that returning 500 would be a default behavior when an exception is raised. but as far as I can see this is not the case.
I have a response transformer back from the destination. But it should not impact as the exception would bypass all that.

**Reply by jonbartels** [reply]
You are correct that an HTTP listener will return 200 for sent and 500 for error by default.

Are you selecting anything in the "respond from" option on the source connector?

Is your destination queueing?

**Reply by cspielmann** [reply]
No queueing / response from the destination
here are the full details:

----

version Mirth Connect Server 3.5.0.8232
Here is my channel setup:

datatype: Json (all inbound/outbound)

**Destination:**  
one single destination _to_GLIMS_CREATE_PATIENT_ that is a channel destination that responds with a json.
Validate response: Yes
Queue messages: Never
Response transformer:  I forced a javascript with _throw ("dummy exception");_

**Source:** 
HTTP_LISTENER (json datatype) / 
Response from: the destination above _to_GLIMS_CREATE_PATIENT_
Response Status code: emtpy
response datatype: text

**postprocessor**
_logger.info("status response = " + responseMap.get('to_GLIMS_CREATE_PATIENT').getStatus());
logger.info("message response = " + responseMap.get('to_GLIMS_CREATE_PATIENT').getMessage());_


**Run:**

with curl: the channel return HTTP/1.1 200 OK
< Date: Wed, 05 Apr 2023 10:13:45 GMT
< Content-Type: application/json; charset=UTF-8
< Content-Length: 0
< Server: Jetty(9.3.14.v20161028)

Logs:

_2023-04-05 12:13:45,601]  ERROR (com.mirth.connect.connectors.vm.VmDispatcher:894): Error executing response transformer for channel GLIMS_CREATE_PATIENT_LISTENER (8a4b510d-d369-4092-b7a4-ba2981aa6a0a) on destination to_GLIMS_CREATE_PATIENT.
com.mirth.connect.donkey.server.channel.components.ResponseTransformerException: CHANNEL:
	GLIMS_CREATE_PATIENT_LISTENERCONNECTOR:
	to_GLIMS_CREATE_PATIENTSCRIPT SOURCE:
	responseSOURCE CODE:
	1167: 1168: /**1169: According to the status of the json received from the destination, we force with an error or not.1170: */1171: 1172: **throw ("dummy exception")**; // TO_REMOVE1173: 1174: if (msg['status']!=='ok') {1175: 
	//responseErrorMessage=msg['error']; 1176: 
	responseStatus=ERROR;LINE NUMBER:
	1172DETAILS:
	dummy exception
	at b5041866-446f-45f7-b0e7-28a6a4f1c9a0:1172 (doTransform)
	at b5041866-446f-45f7-b0e7-28a6a4f1c9a0:1199 (doScript)
	at b5041866-446f-45f7-b0e7-28a6a4f1c9a0:1201
	at com.mirth.connect.server.transformers.JavaScriptResponseTransformer$ResponseTransformerTask.doCall(JavaScriptResponseTransformer.java:167)
	at com.mirth.connect.server.transformers.JavaScriptResponseTransformer$ResponseTransformerTask.doCall(JavaScriptResponseTransformer.java:118)
	at com.mirth.connect.server.util.javascript.JavaScriptTask.call(JavaScriptTask.java:113)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)_

postprocessor exception as well: 
_2023-04-05 12:13:45,734]  ERROR (com.mirth.connect.server.util.javascript.JavaScriptUtil:522): Error executing Postprocessor script from channel: 8a4b510d-d369-4092-b7a4-ba2981aa6a0a
com.mirth.connect.server.MirthJavascriptTransformerException: SOURCE CODE:
	469: 470: /**471: in order to force the HTTP error code of the source HTTP_LISTENER,472: we look at the result of the message from the destination and force an error473: */474: logger.info("status response = " + responseMap.get('to_GLIMS_CREATE_PATIENT').getStatus());475: logger.info("message response = " + responseMap.get('to_GLIMS_CREATE_PATIENT').getMessage());476: 477: //if (msg['status']!=='ok') {478: //
	responseMap.put('to_GLIMS_CREATE_PATIENT', ResponseFactory.getErrorResponse('message'))LINE NUMBER:
	474DETAILS:_


MIRTH Dashboard / Channel messages:
Source: TRANSFORMED  / Errors: Postprocessor
to_GLIMS_CREATE_PATIENT: ERROR / Errors: processing

**Reply by jonbartels** [reply]
Hmmm. Those details look correct. This is strange.

The relevant code for 3.5.0 is https://github.com/nextgenhealthcare/connect/blob/5c97717dcd4afe01b7a8112b7b7bc7291d706937/server/src/com/mirth/connect/connectors/http/HttpReceiver.java#L413

**Reply by cspielmann** [reply]
Yes. this does not seem to work as expected.

I did a very simple test now:

When simple channel with source HTTP_LISTENER and as destination a javascript writer that returns a empty { }. 
then in the response transformer I raise an exception.
then I get still a code 200 when I call the http endpoint

It is very annoying that I cannot raise an error 500 when i have a javascript exception

**Reply by jonbartels** [reply]
>  **in the response transformer** I raise an exception.

That is a key point. I apologize for not picking up on it sooner.  There is probably some different logic that prevents the exception in the response transformer from altering the message state

As a work around the response transformer can alter message statuses. You can set  `responseStatus = ERROR` in the response transformer. If you set the ERROR status in the response transformer explicitly instead of throwing, does the source return HTTP 500?

**Reply by cspielmann** [reply]
Thank for your reply and your hint.

But I even had a doubt in case the exception is raised into any transformer, then does it raise an error 500.
-> I did a test : I raise an exception into the source transformer and even this does **not** set the error code to 500.
I am puzzled.

**Reply by mhokanson** [reply]
I only have a Mirth 4.2 instance to play with, but I put `"slartibartfast".toInt32()` in a Destination Response Transformer for a channel I tried to set up to match your description. When I send `{}` to the HTTP Listener it returned a 500 with the Source Response set to `Auto-generate (Destinations completed)`.

**Reply by cspielmann** [reply]
I think the point is _Auto-generate (Destinations completed)_. 
If I set this config then it works well.

In summary:
in  3.5.0.8232 when I raise an exception in a source transformer, and I have set my HTTP_LISTENER as taking a response from a specific destination, then there is no HTTP code 500 returned. (This destination is not reached as the exception is raised before).

Requirement:
I need to actually get the response message/payload from a specific destination as I have more than one in my channel.

Workaround:
I set _response from postprocessor_
in the postprocessor javascript, I take explicitly the response from the chosen destination and return it.

```
//This script executes once after a message has been processed
//Responses returned from here will be stored as "Postprocessor" in the response map
return responseMap.get('to_transform_esante_eprescription_and_send_to_GLIMS');
```
